### PR TITLE
Use get_conn_flattened for our local energy expectation kernels

### PR DESCRIPTION
**Pull request by PhilipVinc**
**Please give me feedback**

Right now local kernels use `get_conn_padded` which pad the number of connected elements per sample to the maximum number of connected elements among all the samples.

This leads to some wasted compute power.

This PR improves the situation by using `get_conn_flattened` to get the connected elements on discrete operators.
The downside is that the local-energy code becomes considerably more convoluted. The upside is that this gives a net x% speedup where x is the number of 0 matrix elements you have in your operators.

While for Ising and simple hamiltonians this gives no bonus (and in fact, yields a small ~1-5% slowdown in Ising 20 spins, vanishingly small on 60 spins), this gives me a 20% speedup in some dissipative things I'm running.
And I guess @jwnys might be interested too?

Still WIP, but I'd like some feedback on this 

Note: this code path is only taken if chunking is disabled. I still have to update the chunked code.

### CODE REVIEW

File: netket/vqs/mc/kernels.py [hunk 1]
```diff
@@ -70,39 +104,44 @@ def local_value_kernel_flattened(logpsi: Callable, pars: PyTree, σ: Array, args
...
-    σp, mel = args
+    ρ_σσ_ext = jnp.repeat(ρ_σσ, ns, total_repeat_length=N_conn)
 
-    σ_σp = jax.vmap(lambda σp, σ: jnp.hstack((σp, σ)), in_axes=(0, None))(σp, σ)
-    σ_σ = jnp.hstack((σ, σ))
-    return jnp.sum(mel * jnp.exp(logpsi(pars, σ_σp) - logpsi(pars, σ_σ)))
+    delta_ρ = mels * jnp.exp(ρ_ση - ρ_σσ_ext)
+    indices = jnp.repeat(jnp.arange(N_samples), ns, total_repeat_length=N_conn)
+
+    e_loc = jax.ops.segment_sum(
```
Comment (gcarleo) [inline netket/vqs/mc/kernels.py:140]: maybe can you write a comment on what these lines are doing?

Comment (PhilipVinc) [reply netket/vqs/mc/kernels.py:140]: (This is essentially equivalent to the old C++ code that used a for loop, but is written in a jax-compatible way, which makes it rather... unreadable.)

(This PR is still WIP because I'm hoping there's a better way to implement this...)

(If we end up deciding to merge this or some future iteration, I'll of course document it. But for now I'd like some feedback on the performance gain from people like @jwnys who are working with fermonic models which have lots of zeros...)

**Comment by PhilipVinc** [comment]
A particularly important thing for this PR would be the padding strategy: right now if the total number of connected elements changes, this recompiles everything. So it must be padded.

What strategy should we use for padding? 

The standard padding to next power of two is too pessimistic.
I was thinking that a possible option might be to pad to the next multiple of `N_samples`, for example...

**Comment by gcarleo** [comment]
Yes I think multiples of Nsamples is a good idea...powers of two indeed
sounds really too much! The ideal solution would be to run a few find conn
on random x and measure the optimal padding, in a sort of autotuning
mechanism. But sure, it's involved... Do I understand correctly that jax
will eventually avoid recompiling when shapes are not changed but only
sizes? If so, I wouldn't overthink this since it will be eventually removed

**Comment by PhilipVinc** [comment]
> The ideal solution would be to run a few find conn on random x and measure the optimal padding, in a sort of autotuning mechanism

yeah, I was thinking about this. We could add an optional function to operators `estimate_optimal_padding` that does just that, and is used in some cases.
Though implementing an heuristic that works well in general won't be easy.

> Do I understand correctly that jax will eventually avoid recompiling when shapes are not changed but only sizes? If so, I wouldn't overthink this since it will be eventually removed

It's a mid-term priority for them.  From what I understand the idea is that you will be able to declare a dimension as *dynamic*, and then this won't force recompile.

So yeah, it will address this problem.

They had a working implementation but it had some limitation (don't recall what) and they scraped it, and now they are rebuilding it on top of some new LLVM dialect... So I have no idea when this will arrive...

**Comment by PhilipVinc** [comment]
closing this for now...
